### ISSUE TITLE
Ambiguous semaphore state cornercase

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#1501]
Semaphore signal specification says this:

> The first synchronization scope includes every command submitted in the same batch. In the case of `vkQueueSubmit2KHR`, the first synchronization scope is limited to the pipeline stage specified by `VkSemaphoreSubmitInfoKHR::stage`.
> [...]
>The second synchronization scope includes only the semaphore signal operation.

This says semaphore is signaled somewhere undetermined, but later than or at `stageMask`. Actually worse. It kinda says the signal is outside the queue\pipeline stage system entirely.

`VkSemaphoreSubmitInfoKHR::stageMask` is specified as:

> `stageMask` is a `VkPipelineStageFlags2KHR` mask of pipeline stages at which the semaphore either waits or is signaled.

I could interpret this as signal being a queue op with pipeline `VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT` &rArr; `stageMask` &rArr; `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT`, and that signal happening exactly at `stageMask`. Which is somewhat subtly contrary to the previous definition above.

---

Now, where it could make difference is `vkGetSemaphoreCounterValue`. Before the signal op, it has the old value, and after signal it has the new value, right?

So, if I do:

```
vkCmdStuff();
vkQueueSubmit( signal=bad_semaphore, stageMask=STAGE_COOL, value=42 );
vkCmdSetEvent( helper_event, srcStage=STAGE_COOL ); // assume it is submitted
while( vkGetEventStatus(helper_event) != VK_EVENT_SET ) [[pass]];
vkGetSemaphoreCounterValue( bad_semaphore, &current_value );
```

Now, if I use the first spec quote, I would expect the `current_value` to be indeterminable (could be either the old or the new value). But if I use the second spec quote, I would expect `current_value` to always be 42 (new value).

**Comment by HansKristian-Work** [comment]
Ok, I think I understand the confusion here. You're asserting that signalling a semaphore itself is an operation which happens in stageMask, and therefore vkCmdSetEvent somehow ends up waiting for the semaphore signal operation to complete? I don't think this is intended, and I see why the wording "is signaled" is confusing. I'll try to reword it.

**Comment by krOoze** [comment]
In the case of signal, it could obviously happen later than `stageMask`.

The question is generally if it is part of the pipeline stage system. I.e. if I can infer the state of a semaphore after a dependency with `VK_PIPELINE_STAGE_ALL_COMMANDS_BIT`.

**Comment by oddhack** [comment]
This should be fixed in the 1.2.180 spec update.
